addons/project/tests/test_project_milestone.py

```diff
@@ -36,20 +39,56 @@ def test_milestones_settings_change(self):
-        """ Test when a task is linked to a milestone and when we change its project the milestone is removed
+        """ Test when a task is linked to a milestone and when we change its project the milestone is removed (and
+            we fallback on the parent milestone if it belongs to the same project)
 
             Test Case:
             =========
             1) Set a milestone on the task
             2) Change the project of that task
-            3) Check no milestone is linked to the task
+            3) Check no milestone is linked to the task (or the one of its parent is used if relevant)
         """
-        self.task_1.milestone_id = self.milestone
-        self.assertEqual(self.task_1.milestone_id, self.milestone)
+        # A. No parent task
+        self.task_1.milestone_id = self.milestone_pigs
+        self.assertEqual(self.task_1.milestone_id, self.milestone_pigs)
+
 
         self.task_1.project_id = self.project_goats
         self.assertFalse(self.task_1.milestone_id, 'No milestone should be linked to the task since its project has changed')
 
+        # B. Parent task with no milestone set
+        self.task_2 = self.env['project.task'].with_context({'mail_create_nolog': True}).create({
+            'name': 'Chikd MilestoneTask',
+            'user_ids': self.user_projectmanager,
+            'project_id': self.project_pigs.id,
+            'parent_id': self.task_1.id,
+            'milestone_id': self.milestone_pigs.id,
+        })
+        self.assertEqual(self.task_2.milestone_id, self.milestone_pigs)
+
+        self.task_2.project_id = self.project_goats
+        self.assertFalse(self.task_2.milestone_id, 'No milestone should be linked to the task since its project has changed and its parent task has no milestone')
+
+        # C. Parent task with a milestone set but on a different project
+        self.task_1.milestone_id = self.milestone_pigs
+        self.task_1.project_id = self.project_pigs
+        self.task_2.project_id = self.project_pigs
+        self.task_2.milestone_id = self.milestone_pigs
+        self.assertEqual(self.task_2.milestone_id, self.milestone_pigs)
+
+        self.task_2.project_id = self.project_goats
+        self.assertFalse(self.task_2.milestone_id, 'No milestone should be linked to the task since its project has changed and its parent task belongs to another project')
+
+        # D. Parent task with a milestone set but on the same project
+        self.task_1.milestone_id = self.milestone_goats
+        self.task_1.project_id = self.project_goats
+        self.task_2.project_id = self.project_pigs
+        self.task_2.milestone_id = self.milestone_pigs
+        self.assertEqual(self.task_2.milestone_id, self.milestone_pigs)
+
+        self.task_2.project_id = self.project_goats
+        self.assertFalse(self.task_2.milestone_id, 'The milestone of the task should be replaced by the one of its parent task as they now belong to the same project')
```

Comment:
I'm confused, shouldn't here the ``milestone_id`` be ``self.milestone_goats`` ? the same as the parent's task?
This is what I gathered from here: "we fallback on the parent milestone if it belongs to the same project"

Reply:
> I'm confused, shouldn't here the `milestone_id` be `self.milestone_goats` ? the same as the parent's task? This is what I gathered from here: "we fallback on the parent milestone if it belongs to the same project"

You're :100: right, this test was not doing what it tells, I've changed it :)